src/3d/qgs3daxis.h

```diff
@@ -0,0 +1,218 @@
+     * @param parent3DScene root entity to set as parent
+     * @param camera camera controller used to track camera movements
+     * @param map 3D map settings
+     */
+    Qgs3DAxis( Qt3DExtras::Qt3DWindow *parentWindow,  Qt3DCore::QEntity *parent3DScene, QgsCameraController *camera, const Qgs3DMapSettings *map );
+
+    /**
+     * \brief The Axis enum
+     */
+    enum class Axis
+    {
+      X = 1,
+      Y = 2,
+      Z = 3
+    };
+    Q_ENUM( Axis )
+
+    /**
+     * \brief The AxisViewportPosition enum
+     */
+    enum class AxisViewportPosition
+    {
+      //! top or left
+      Begin = 1,
+      Middle = 2,
+      //! bottom or right
+      End = 3
+    };
+    Q_ENUM( AxisViewportPosition )
+
+    /**
+     * \brief The Mode enum
+     */
+    enum class Mode
+    {
+      //! disabled
+      Off = 1,
+      //! CRS specific.
+      Crs = 2,
+      //! Cube with label
+      Cube = 3
+    };
+    Q_ENUM( Mode )
+
+    /**
+     * \brief set axis representation mode
+     * \param axisMode new node
+     */
+    void setMode( Qgs3DAxis::Mode axisMode );
+
+    /**
+     * \brief set axis viewport position parameters
+     * \param axisViewportSize height/width size in pixel
+     * @param axisViewportVertPos start vertical position
+     * @param axisViewportHorizPos start horizontal position
+     */
+    void setAxisViewportPosition( int axisViewportSize, AxisViewportPosition axisViewportVertPos, AxisViewportPosition axisViewportHorizPos );
+
+    /**
+     * \brief get axis viewport size
+     */
+    int axisViewportSize() const { return mAxisViewportSize;}
+
+    /**
+     * \brief get axis viewport horizontal position
+     */
+    AxisViewportPosition axisViewportHorizontalPosition() const { return mAxisViewportHorizPos;}
+
+    /**
+     * \brief get axis viewport vertical position
+     */
+    AxisViewportPosition axisViewportVerticalPosition() const { return mAxisViewportVertPos;}
+
+    /**
+     * \brief project a 3D position from sourceCamera (in sourceViewport) to a 2D position for destCamera (in destViewport). destCamera and the destViewport act as a billboarding layer. The labels displayed by this process will always face the camera.
+     *
+     * \param sourcePos 3D label coordinates
+     * @param sourceCamera main view camera
+     * @param sourceViewport main viewport
+     * @param destCamera billboarding camera
+     * @param destViewport billboarding viewport
+     * @param destSize main qt3d window size
+     * @return
+     */
+    QVector3D from3dTo2dLabelPosition( const QVector3D &sourcePos,
+                                       Qt3DRender::QCamera *sourceCamera, Qt3DRender::QViewport *sourceViewport,
+                                       Qt3DRender::QCamera *destCamera, Qt3DRender::QViewport *destViewport,
+                                       const QSize &destSize );
+
+  private:
+    void createAxisScene();
+    void createAxis( const Axis &axis );
```

Comment:
I think it is better if a function with the prefix returned the thing it created. That will make the code easier to understand.

Reply:
This function creates different objects per axis and then can not return anything.